src/java/org/apache/cassandra/index/sai/iterators/HnswIntersectionIterator.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.iterators;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+
+/**
+ * The HnswIntersectionIterator is an implementation of the KeyRangeIterator
+ * that handles intersection operations between normal SAI predicates and Hierarchical
+ * Navigable Small World (HNSW) graphs.
+ *
+ * HNSW is special because it returns results in similarity order, not in key order,
+ * and because it's really more of an order by than a where clause; HNSW doesn't
+ * "eliminate" any keys, they will all be in the results if you iterate far down enough.
+ *
+ * HnswIntersectionIterator performs what the Pinterest guys call Post Filtering:
+ * we perform the HNSW search, then go through the keys from the other predicates
+ * looking for matches.  If we don't get enough matches, we expand the HNSW search
+ * window and perform the search over again.
+ *
+ * To attempt to minimize the passes, we double the HNSW window with each iteration.
+ */
+public class HnswIntersectionIterator extends KeyRangeIterator
+{
+    private final Builder builder;
```

Comment:
nit: We don't need to store this, we only need to store the `limit`.

Reply:
I left the `otherSuppliers` in the builder too